include/nbl/builtin/hlsl/emulated_float64_t.hlsl

```diff
@@ -0,0 +1,499 @@
+#else
+                        return createPreserveBitPattern(impl::propagateFloat64NaN(data, rhs.data));
+#endif
+                    if ((uint64_t(lhsBiasedExp) | lhsMantissa) == 0u)
+                        return createPreserveBitPattern(ieee754::traits<float64_t>::quietNaN);
+
+                    return createPreserveBitPattern(sign | ieee754::traits<float64_t>::exponentMask);
+                }
+                if (lhsBiasedExp == 0)
+                {
+                    if (lhsMantissa == 0u)
+                        return createPreserveBitPattern(sign);
+                    impl::normalizeFloat64Subnormal(lhsMantissa, lhsBiasedExp, lhsMantissa);
+                }
+                if (rhsBiasedExp == 0)
+                {
+                    if (rhsMantissa == 0u)
+                        return createPreserveBitPattern(sign);
+                    impl::normalizeFloat64Subnormal(rhsMantissa, rhsBiasedExp, rhsMantissa);
+                }
+            }
+            
+            const uint64_t hi_l = (lhsMantissa >> 21) | (1ull << 31);
+            const uint64_t lo_l = lhsMantissa & ((1ull << 21) - 1);
+            const uint64_t hi_r = (rhsMantissa >> 21) | (1ull << 31);
+            const uint64_t lo_r = rhsMantissa & ((1ull << 21) - 1);
+
+            //const uint64_t RoundToNearest = (1ull << 31) - 1;
+            uint64_t newPseudoMantissa = ((hi_l * hi_r) >> 10) + ((hi_l * lo_r + lo_l * hi_r/* + RoundToNearest*/) >> 31);
+
+            if (newPseudoMantissa & (0x1ull << 53))
+            {
+                newPseudoMantissa >>= 1;
+                ++exp;
+            }
+            newPseudoMantissa &= (ieee754::traits<float64_t>::mantissaMask);
+
+            return createPreserveBitPattern(impl::assembleFloat64(sign, uint64_t(exp) << ieee754::traits<float64_t>::mantissaBitCnt, newPseudoMantissa));
+        }
+
+        emulated_float64_t operator/(const emulated_float64_t rhs)
+        {
+            // TODO: maybe add function to extract real mantissa
+            const uint64_t lhsRealMantissa = (ieee754::extractMantissa(data) | (1ull << ieee754::traits<float64_t>::mantissaBitCnt));
+            const uint64_t rhsRealMantissa = ieee754::extractMantissa(rhs.data) | (1ull << ieee754::traits<float64_t>::mantissaBitCnt);
+            
+            const uint64_t sign = (data ^ rhs.data) & ieee754::traits<float64_t>::signMask;
+            int exp = ieee754::extractExponent(data) - ieee754::extractExponent(rhs.data) + ieee754::traits<float64_t>::exponentBias;
+
+            uint64_t2 lhsMantissaShifted = impl::shiftMantissaLeftBy53(lhsRealMantissa);
+            uint64_t mantissa = impl::divmod128by64(lhsMantissaShifted.x, lhsMantissaShifted.y, rhsRealMantissa);
+
+            if (mantissa & (0x1ull << 53))
+            {
+                ++exp;
+            }
+            else
+            {
+                mantissa >>= 1;
+            }
+
+            mantissa &= ieee754::traits<float64_t>::mantissaMask;

```

Comment:
I'll take a look at this soon if the reciprocal version doesn't turn out to work and be faster